common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -509,6 +509,10 @@ private CollationSpecUTF8(
 
       private static int collationNameToId(String originalName, String collationName)
           throws SparkException {
+        // Have a check for UTF8_BINARY collation to early-out and not introduce any regression.
+        if (UTF8_BINARY_COLLATION.collationName.equals(collationName)) {
```

Comment:
You could avoid double comparison to `UTF8_BINARY_COLLATION` if you modify the `if` below like:
```java
        if (collationName.startsWith(UTF8_BINARY_COLLATION.collationName)) {
          if (collationName.length() == UTF8_BINARY_COLLATION.collationName.length()) {
            return UTF8_BINARY_COLLATION_ID;
          }
```

Reply:
Done. Added special case for `UTF8_LCASE` as well.

Reply:
FWI, according to the current (misleading) benchmarks, this approach yields ~40% slower results.